skada/deep/base.py

```diff
@@ -0,0 +1,392 @@
+                negative_iter = iter(negative_sampler)
+                neg_idx = self.negative_indices[next(negative_iter)]
+            yield pos_idx
+            yield neg_idx
+
+    def __len__(self):
+        return 2 * self.num_samples
+
+
+class DomainBalancedDataLoader(DataLoader):
+    """Domain balanced data loader
+
+    A data loader to have as much as source and target in the batch.
+
+    Parameters
+    ----------
+    dataset : torch dataset
+        The dataset to sample from.
+    """
+
+    def __init__(
+        self,
+        dataset,
+        batch_size,
+        shuffle=False,
+        sampler=None,
+        batch_sampler=None,
+        num_workers=0,
+        collate_fn=None,
+        pin_memory=False,
+        drop_last=False,
+        timeout=0,
+        worker_init_fn=None,
+        multiprocessing_context=None,
+    ):
+        sampler = DomainBalancedSampler(dataset)
+        super().__init__(
+            dataset,
+            2 * batch_size,
+            shuffle,
+            sampler,
+            batch_sampler,
+            num_workers,
+            collate_fn,
+            pin_memory,
+            drop_last,
+            timeout,
+            worker_init_fn,
+            multiprocessing_context,
+        )
+
+
+class DomainAwareModule(torch.nn.Module):
+    """Domain aware module
+
+    A domain aware module allowing to separe the source and target and
+    compute their respective prediction and feaures.
+
+    Parameters
+    ----------
+    module : torch module (class or instance)
+        A PyTorch :class:`~torch.nn.Module`.
+    layer_name : str
+        The name of the module's layer whose outputs are
+        collected during the training for the adaptation.
+    domain_classifier : torch module, default=None
+        A PyTorch :class:`~torch.nn.Module` used to classify the
+        domain. Could be None.
+    """
+
+    def __init__(self, module, layer_name, domain_classifier=None):
+        super(DomainAwareModule, self).__init__()
+        self.module_ = module
+        self.domain_classifier_ = domain_classifier
+        self.layer_name = layer_name
+        self.intermediate_layers = {}
+        _register_forwards_hook(
+            self.module_, self.intermediate_layers, [self.layer_name]
+        )
+
+    def forward(self, X, sample_domain=None, is_fit=False, return_features=False):
+        if is_fit:
+            X_s = X[sample_domain > 0]
+            X_t = X[sample_domain < 0]
+            # predict
+            y_pred_s = self.module_(X_s)
+            features_s = self.intermediate_layers[self.layer_name]
+            y_pred_t = self.module_(X_t)
+            features_t = self.intermediate_layers[self.layer_name]
+
+            if self.domain_classifier_ is not None:
+                domain_pred_s = self.domain_classifier_(features_s)
+                domain_pred_t = self.domain_classifier_(features_t)
+                domain_pred = torch.empty((len(sample_domain)))
```

Comment:
I guess this tensor should be allocated on the same device where the module is, right?

Reply:
Let's just open an issue for now.